File: AUGUST 24/20 August/Burning Tree gfg.cpp
class Solution {
  public:
    int minTime(Node* root, int target) 
    {
        // Your code goes here
        
        unordered_map<int, vector<int>> adjList;
        queue<Node*> nodes;
        nodes.push(root);
        while(!nodes.empty())
        {
            int N = nodes.size();
            for(int i=0; i<N; i++)
            {
                Node* node = nodes.front();
                nodes.pop();
                if(node->left)
                {
                    adjList[node->data].push_back(node->left->data);
                    adjList[node->left->data].push_back(node->data);
                    nodes.push(node->left);
                }
                if(node->right)
                {
                    adjList[node->data].push_back(node->right->data);
                    adjList[node->right->data].push_back(node->data);
                    nodes.push(node->right);
                }
                
            }
        }

        
        unordered_map<int,bool> visited;
        queue<pair<int,int>> q;
        q.push({target,0});
        visited[target] = true;
        int minimumTime = 0;
        
        while(!q.empty() )
        {
            pair<int,int> p = q.front();
            q.pop();
            
            minimumTime = max(minimumTime, p.second);
            for(int next:adjList[p.first])
            {
                if(!visited[next])
                {
                    q.push({next, p.second+1});
                    visited[next] = true;
                }
            }
            // cout<<endl;
            
        }
        return minimumTime;
    }
};

File: MAY 2025/20 May/Burning Tree gfg.java
class Solution {
    static Map<Node, Node> parentMap = new HashMap<>();
    private static Node _targetNode;
    private static int _target;
    
    static void findParent(Node root, Node parent)
    {
        if(root==null)
            return;
        if(root.data == _target)
            _targetNode = root;
        parentMap.put(root, parent);
        findParent(root.left, root);
        findParent(root.right, root);
    }
    
    private static int burnTree()
    {
        Queue<Node> q = new LinkedList<>();
        q.offer(_targetNode);
        boolean[] visited = new boolean[100005];
        visited[_targetNode.data] = true;
        int time = -1;
        while(!q.isEmpty())
        {
            int qsize = q.size();
            for(int i=0; i<qsize; i++)
            {
                Node node = q.poll();
                
                // System.out.print(node.data+" ");
                
                Node parent = parentMap.get(node);
                if(parent != null && !visited[parent.data])
                {
                    q.offer(parent);
                    visited[parent.data] = true;
                }
                
                if(node.left != null && !visited[node.left.data])
                {
                    q.offer(node.left);
                    visited[node.left.data] = true;
                }
                if(node.right != null &&  !visited[node.right.data])
                {
                    q.offer(node.right);
                    visited[node.right.data] = true;
                }
            }
            time++;
            // System.out.println(" time "+time);
        }
        return time;
    }
    public static int minTime(Node root, int target) {
        // code here
        _target = target;
        findParent(root, null);
        return burnTree();
    }
}
